File: zb32_clojure/src/zb32.clj
(ns zb32
    (:require [clojure.tools.cli :as cli])
    (:gen-class))


(def alphabet "ybndrfg8ejkmcpqxot1uwisza345h769")

(defn generate-zb32-string [rng length]
    (apply
        str
        (repeatedly length #(get alphabet (.nextInt rng 32)))))

(def param-spec
    [["-s" "--seed SEED" "Use a specific random seed"
      :parse-fn parse-long]
     ["-l" "--length LENGTH" "Length of each generated string"
      :parse-fn parse-long
      :default 16]
     ["-c" "--count COUNT" "Count of strings to generate"
      :parse-fn parse-long
      :default 1]
     ["-h" "--help"]])

(defn -main [& argv]
    (def args (cli/parse-opts argv param-spec))
    (def options (args :options))
    (if (options :help)
        (do
            (println "Z-Base-32 random string generator\n")
            (println "usage: zb32 [options]")
            (println (args :summary))
            (System/exit 0)))

    (def rng
        (if (options :seed)
            (new java.util.Random (options :seed))
            (new java.util.Random)))

    (dotimes [_ (options :count)]
        (println (generate-zb32-string rng (options :length))))
)

File: zb32_golang/zb32.go
package main

import (
    "github.com/docopt/docopt-go"
    "math/rand"
    "fmt"
    "os"
    "strings"
    "strconv"
    "time"
)

const ALPHABET = "ybndrfg8ejkmcpqxot1uwisza345h769"

func generate_zb32_string(length uint8) string {
    chars := make([]string, length)

    var i uint8 = 0
    for ; i < length; i++ {
        chars[i] = string(ALPHABET[rand.Intn(32)])
    }

    return strings.Join(chars, "")
}

func main() {
    usage := `Z-Base-32 random string generator.

Usage:
  zb32 [--seed SEED] [--length LENGTH] [--count COUNT]
  zb32 -h | --help

Options:
  -h --help           Show this screen.
  -s --seed SEED      Use a specific random seed
  -l --length LENGTH  Length of each generated string [default: 16]
  -c --count COUNT    Count of strings to generate [default: 1]`

    arguments, _ := docopt.Parse(usage, nil, true, "ZB32 1.0", false)

    if arguments["--help"].(bool) {
        fmt.Println(usage)
        os.Exit(1)
    }

    if arguments["--seed"] != nil {
        seed, _ := strconv.ParseInt(arguments["--seed"].(string), 0, 64)
        rand.Seed(seed)
    } else {
        rand.Seed(time.Now().UnixNano())
    }


    tempCount, _ := strconv.ParseInt(arguments["--count"].(string), 0, 32)
    var count int = int(tempCount)

    tempLength, _ := strconv.ParseUint(arguments["--length"].(string), 0, 8)
    var length uint8 = uint8(tempLength)

    for i := 0; i < count; i++ {
        fmt.Println(generate_zb32_string(length))
    }

    os.Exit(0)
}

